### ISSUE TITLE
azurerm_role_assignment.role_definition_id changes upon every apply

### ISSUE THREAD

**Issue by unique-dominik** [hashicorp/terraform-provider-azurerm#25328]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.7.0

### AzureRM Provider Version

3.84.0

### Affected Resource(s)/Data Source(s)

azurerm_role_assignment

### Terraform Configuration Files

```hcl
resource "azurerm_role_assignment" "sub_iac_principal" {
  count = length(local.all_subscriptions)

  scope              = local.all_subscriptions[count.index].id
  role_definition_id = azurerm_role_definition.sub_iac_principal.role_definition_resource_id
  principal_id       = azuread_service_principal.iac_principal.id
}
```

### Debug Output/Panic Output

```shell
-/+ resource "azurerm_role_assignment" "sub_iac_principal" {
      ~ id                               = "/subscriptions/57b03935-48e1-4f0e-8d27-1aca35c8286a/providers/Microsoft.Authorization/roleAssignments/96586121-b7f7-4383-af30-03fd71e0f0ac" -> (known after apply)
      ~ name                             = "e40fba9d-1b06-4791-8cdb-b3523feed0b8" -> (known after apply)
      ~ principal_type                   = "ServicePrincipal" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/abf4ef76-5ae9-43b5-a4b2-b0c9542019ea/providers/Microsoft.Authorization/roleDefinitions/fca28c48-d245-43bb-b04c-760750f7bf82" -> "/providers/Microsoft.Authorization/roleDefinitions/48c4a17e-9401-4783-9836-0fcca22521f2" # forces replacement
      ~ role_definition_name             = "Infrastructure as Code Service Principal" -> (known after apply)
      + skip_service_principal_aad_check = (known after apply)
        # (2 unchanged attributes hidden)
    }


_Note:_ The UUID's are exchanged for randoms.
```

### Expected Behaviour

Re-Applying the same change twice should result in no diff. 

### Actual Behaviour

Yet this diff appears every time basically with the pattern:

```hcl
/subscriptions/<uuid1>/providers/Microsoft.Authorization/roleDefinitions/<uuid2>" 

-> 

"/providers/Microsoft.Authorization/roleDefinitions/<uuid2>"
```

### Important Factoids

I am 20000% sure that this issue was not the case 2-3 weeks ago as then I ran this plan dozens of times with no `diff` multiple times per day over weeks.

**Comment by ziyeqf** [comment]
Hi @unique-dominik, this is a known and fixed issue, the fixing PR is https://github.com/hashicorp/terraform-provider-azurerm/pull/24266 and released on v3.91.0 version, could you please upgrade your provider version and try again?

Thanks

**Comment by magodo** [comment]
@unique-dominik There are two issues around this:

1. Why the first diff occurs
2. Why the plan diff after apply occurs

Begin with the 2nd one: The role assignment service will conditionally *modify* the `role_definition_id` depending on the `azurerm_role_assignment.scope`.

E.g. given:

- The `azurerm_role_assignment.role_definition_id` as a tenant level one (e.g. `/providers/Microsoft.Authorization/roleDefinitions/123`)
- The `azurerm_role_assignment.scope` specified to a subscription (e.g. `/subscriptions/456`)

Then the returned `azurerm_role_assignment.role_definition_id` will be changed to `/subscriptions/456/providers/Microsoft.Authorization/roleDefinitions/123`, which aligns with the scope of the current `azurerm_role_assignment`'s id. This is the API behavior all the time.

So to workaround your issue, you can modify your TF config to prepend with the subscription id, like below:

```
role_definition_id = "${local.all_subscriptions[count.index].id}${azurerm_role_definition.sub_iac_principal.role_definition_}"resource_id
```

This is what we are doing for the test:

https://github.com/hashicorp/terraform-provider-azurerm/blob/91352c2a9362d040f5153b42a790f5f3f853bd0e/internal/services/authorization/role_assignment_resource_test.go#L292-L296

Besides, regarding the source of the `azurerm_role_assignment.role_definition_id`, i.e. the `azurerm_role_definition.role_definition_resource_id`, its format depends on how we `GET` it. As a role definition can have multiple assignable scopes, it means you can `GET` it via different prefix scopes. Depending on how you `GET` it (e.g. with only the `scope` prefix, or with one of the `assignable_scopes`), it makes the result response's id has the same scope prefix.

Back to the first issue, as the provider version didn't change, and also the API behavior shall remain still. I suspect you've some out-of-band change on the `azurerm_role_definition` resource (e.g. change of the scope).

**Comment by unique-dominik** [comment]
Excellent feedback thank you so much! You mentioning `scope` rang a bell. I will check next week but it might be that we changed a scope, applied but then did not re-plan or re-apply until yesterday 👀 
Coming back next week…!

**Comment by unique-dominik** [comment]
@magodo You are (what I expected) correct, we indeed changed the scope but already 5 weeks (now 6 weeks) ago so where no-one did plan or apply. So yeah, I'll dig that out and fix it.

For everyone else seeing the diff as we do, we made this change:
<img width="1060" alt="image" src="https://github.com/hashicorp/terraform-provider-azurerm/assets/103249640/bb79705e-908f-4cee-96d7-545cf5ce23bb">

and naturally applied it and went home happily ever after. Once we replanned weeks later the diff shows up.

Thank you @magodo ❤️
